crates/sui-framework/packages/sui-framework/sources/crypto/poseidon.move

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Module which defines instances of the poseidon hash functions.
+module sui::poseidon {
+    /// Error if any of the inputs are larger than or equal to the BN254 field size.
+    const ENonCanonicalInput: u64 = 0;
+
+    /// Error if more than 32 inputs are provided.
+    const ETooManyInputs: u64 = 1;
+
+    /// @param data: Vector of BN254 field elements to hash.
+    ///
+    /// Hash the inputs using poseidon_bn254 and returns a BN254 field element.
+    ///
+    /// The number of inputs cannot exceed 32 and each element has to be a BN254
+    /// field element in canonical representation so it must be smaller than
+    /// the BN254 scalar field size which is
+    /// 21888242871839275222246405745257275088548364400416034343698204186575808495617.
```

Comment:
Can we add some explanation on the serialization format of the inputs (endianness and co). Mentioning this because there is not a widely accepted standard necessarily for Poseidon that we adhere to.

Additionally, let's have a pointer re the discussion we had around hashing arbitrary bytes, cause although I can understand how can someone provide inputs from offchain serialization at entry functions, it won't be clear how someone can use Poseidon inside a smart contract (ie hash some bytes produced by a smart contract)

Reply:
As the API is here, there's actually no issue with serialization because the input is a move type, u256. This was chosen exactly to avoid getting into a non-standard API and discuss field element representation. A contract may use u256's `from_le_bytes` method if it wants to hash binary data but should take into account that this will give an error if the 32 bytes as a little-endian encoded integer gives in a non-canonical representation.

A generic function to hash bytes could either chunk in 31 bytes. Or it could consider the binary input as a very large integer, write it in base r (the field size) and then hash the digits. The latter will be the most optimal approach, but perhaps not easy to explain to builders.